src/sage/rings/species.py

```diff
@@ -0,0 +1,1073 @@
+
+    def __eq__(self, other):
+        r"""
+        Two atomic species are equal if the underlying groups are conjugate,
+        and their domain partitions are equal under the conjugating element.
+
+        TESTS::
+
+            sage: At = AtomicSpecies(2)
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: H = PermutationGroup([[(1,2,3,4),(5,6),(7,8),(9,10)]]); H
+            Permutation Group with generators [(1,2,3,4)(5,6)(7,8)(9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+            sage: B = At(H, {1: [1,2,3,4], 2: [5,6,7,8,9,10]})
+            sage: C = At(G, {1: [1,2,5,6], 2: [3,4,7,8,9,10]})
+            sage: A != B
+            True
+            sage: A == C
+            True
+        """
+        if parent(self) != parent(other):
+            return False
+        # Check if multicardinalities match
+        if self._mc != other._mc:
+            return False
+        # If they do, construct the mapping between the groups
+        selflist, otherlist = [], []
+        for i in range(self.parent()._k):
+            if len(self._dompart[i]) != len(other._dompart[i]):
+                return False
+            selflist.extend(self._dompart[i])
+            otherlist.extend(other._dompart[i])
+        mapping = libgap.MappingPermListList(selflist, otherlist)
+        G = PermutationGroup(gap_group=libgap.ConjugateGroup(self._dis._C, mapping),
+                             domain=self._dis._C.domain())
+        # The conjugated group must be exactly equal to the other group
+        return G == other._dis._C
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: At = AtomicSpecies(2)
+            sage: G = PermutationGroup([[(1,2),(3,4),(5,6),(7,8,9,10)]]); G
+            Permutation Group with generators [(1,2)(3,4)(5,6)(7,8,9,10)]
+            sage: A = At(G, {1: [1,2,3,4], 2: [5,6,7,8,9,10]}); A
+            {[(1,2,3,4)(5,6)(7,8)(9,10)]: ((5, 6, 7, 8), (9, 10, 1, 2, 3, 4))}
+        """
+        return "{" + f"{self._dis}: {self._dompart}" + "}"
+
+class AtomicSpecies(UniqueRepresentation, Parent, ElementCache):
+    def __init__(self, k):
+        r"""
+        Infinite set of `k`-variate atomic species graded by
+        integer vectors of length `k`.
+
+        TESTS::
+
+            sage: At1 = AtomicSpecies(1)
+            sage: At2 = AtomicSpecies(2)
+            sage: TestSuite(At1).run(skip="_test_graded_components")
+            sage: TestSuite(At2).run(skip="_test_graded_components")
+        """
+        category = SetsWithGrading().Infinite()
+        Parent.__init__(self, category=category)
+        ElementCache.__init__(self)
+        self._k = k
+        self._grading_set = IntegerVectors(length=k)
+        self._dis_ctor = ConjugacyClassesOfDirectlyIndecomposableSubgroups()
```

Comment:
This is used only once currently, and makes the code harder to read.  I suggest to inline this, i.e., do
```
        dis_elm = ConjugacyClassesOfDirectlyIndecomposableSubgroups()(G)
```
in `_element_constructor_ `.